File: surface/widgets/compose/components/widgets/domain/WorkflowEditor.kt
// ============================================================
// Clef Surface Compose Widget — WorkflowEditor
//
// Node-graph workflow canvas rendered as a Column with a
// LazyColumn of nodes and their connections. Supports node
// selection, adding/removing nodes, and displays connections
// between nodes as indented sub-items.
//
// Adapts the workflow-editor.widget spec: anatomy (root, canvas,
// nodePalette, configPanel, minimap, toolbar, executeButton),
// states (idle, nodeSelected, configuring, placing, draggingNew,
// executing, executionResult), and connect attributes.
// ============================================================

package clef.surface.compose.components.widgets.domain

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

// --------------- Types ---------------

data class WorkflowEditorNode(
    val id: String,
    val type: String,
    val label: String,
)

data class WorkflowConnection(
    val from: String,
    val to: String,
)

// --------------- Component ---------------

/**
 * Node-graph workflow editor with a LazyColumn of nodes and connections.
 *
 * @param nodes Nodes in the workflow graph.
 * @param connections Connections between nodes.
 * @param selectedNode ID of the currently selected node.
 * @param onSelectNode Callback when a node is selected.
 * @param onAddNode Callback to add a new node.
 * @param onRemoveNode Callback to remove a node by id.
 * @param modifier Modifier for the root layout.
 */
@Composable
fun WorkflowEditor(
    nodes: List<WorkflowEditorNode>,
    connections: List<WorkflowConnection>,
    selectedNode: String? = null,
    onSelectNode: (String) -> Unit = {},
    onAddNode: () -> Unit = {},
    onRemoveNode: (String) -> Unit = {},
    modifier: Modifier = Modifier,
) {
    // Build adjacency
    val adjacency = mutableMapOf<String, MutableList<String>>()
    connections.forEach { conn ->
        adjacency.getOrPut(conn.from) { mutableListOf() }.add(conn.to)
    }
    val nodeLabels = nodes.associate { it.id to it.label }

    Column(modifier = modifier.padding(8.dp)) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(
                text = "Workflow",
                style = MaterialTheme.typography.titleMedium,
                fontWeight = FontWeight.Bold,
            )
            Text(
                text = " (${nodes.size} nodes, ${connections.size} connections)",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }

        Spacer(modifier = Modifier.height(8.dp))

        LazyColumn {
            itemsIndexed(nodes, key = { _, node -> node.id }) { _, node ->
                val isSelected = node.id == selectedNode
                val targets = adjacency[node.id] ?: emptyList()

                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = 2.dp)
                        .clickable { onSelectNode(node.id) },
                    colors = CardDefaults.cardColors(
                        containerColor = if (isSelected)
                            MaterialTheme.colorScheme.primaryContainer
                        else
                            MaterialTheme.colorScheme.surface,
                    ),
                ) {
                    Column(modifier = Modifier.padding(12.dp)) {
                        Row(verticalAlignment = Alignment.CenterVertically) {
                            Text(
                                text = "[${node.type}]",
                                style = MaterialTheme.typography.labelMedium,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                            )
                            Spacer(modifier = Modifier.width(8.dp))
                            Text(
                                text = node.label,
                                style = MaterialTheme.typography.bodyMedium,
                                fontWeight = if (isSelected) FontWeight.Bold else FontWeight.Normal,
                            )
                        }
                        // Connection targets
                        targets.forEach { targetId ->
                            val targetLabel = nodeLabels[targetId] ?: targetId
                            Text(
                                text = "  \u2514\u2500\u2500\u2192 $targetLabel",
                                style = MaterialTheme.typography.bodySmall,
                                color = MaterialTheme.colorScheme.onSurfaceVariant,
                                modifier = Modifier.padding(start = 16.dp, top = 2.dp),
                            )
                        }
                    }
                }
            }
        }

        if (nodes.isEmpty()) {
            Text(
                text = "(empty workflow)",
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(16.dp),
            )
        }

        Spacer(modifier = Modifier.height(8.dp))

        OutlinedButton(
            onClick = onAddNode,
            modifier = Modifier.fillMaxWidth(),
        ) {
            Text(text = "+ Add Node")
        }
    }
}

File: surface/widgets/gtk/components/widgets/domain/WorkflowEditor.ts
// ============================================================
// Clef Surface GTK Widget — WorkflowEditor
//
// Node-graph workflow canvas rendered as a vertical list of
// nodes and their connections. Supports node selection,
// adding/removing nodes, and displays connections between
// nodes as indented sub-items.
//
// Adapts the workflow-editor.widget spec: anatomy (root, canvas,
// nodePalette, configPanel, minimap, toolbar, executeButton),
// states (idle, nodeSelected, configuring, placing, draggingNew,
// executing, executionResult), and connect attributes to GTK4
// rendering.
// ============================================================

import Gtk from 'gi://Gtk?version=4.0';

// --------------- Types ---------------

export interface WorkflowEditorNode { id: string; type: string; label: string; }
export interface WorkflowConnection { from: string; to: string; }

// --------------- Props ---------------

export interface WorkflowEditorProps {
  nodes?: WorkflowEditorNode[];
  connections?: WorkflowConnection[];
  selectedNode?: string | null;
  onSelectNode?: (id: string) => void;
  onAddNode?: () => void;
  onRemoveNode?: (id: string) => void;
}

// --------------- Component ---------------

export function createWorkflowEditor(props: WorkflowEditorProps = {}): Gtk.Widget {
  const { nodes = [], connections = [], selectedNode = null, onSelectNode, onAddNode, onRemoveNode } = props;

  const adjacency = new Map<string, string[]>();
  connections.forEach((conn) => {
    if (!adjacency.has(conn.from)) adjacency.set(conn.from, []);
    adjacency.get(conn.from)!.push(conn.to);
  });
  const nodeLabels = new Map(nodes.map((n) => [n.id, n.label]));

  const box = new Gtk.Box({ orientation: Gtk.Orientation.VERTICAL, spacing: 8 });

  // Header
  const header = new Gtk.Box({ orientation: Gtk.Orientation.HORIZONTAL, spacing: 4 });
  const titleLabel = new Gtk.Label({ label: 'Workflow', xalign: 0 });
  titleLabel.get_style_context().add_class('heading');
  header.append(titleLabel);
  const statsLabel = new Gtk.Label({ label: `(${nodes.length} nodes, ${connections.length} connections)` });
  statsLabel.get_style_context().add_class('dim-label');
  header.append(statsLabel);
  box.append(header);

  // Node list
  nodes.forEach((node) => {
    const isSelected = node.id === selectedNode;
    const targets = adjacency.get(node.id) ?? [];

    const card = new Gtk.Box({ orientation: Gtk.Orientation.VERTICAL, spacing: 4 });
    card.get_style_context().add_class('card');
    if (isSelected) card.get_style_context().add_class('accent');

    const nodeRow = new Gtk.Box({ orientation: Gtk.Orientation.HORIZONTAL, spacing: 8 });
    const typeLabel = new Gtk.Label({ label: `[${node.type}]` });
    typeLabel.get_style_context().add_class('dim-label');
    nodeRow.append(typeLabel);
    nodeRow.append(new Gtk.Label({ label: node.label, xalign: 0, hexpand: true }));
    card.append(nodeRow);

    // Connection targets
    targets.forEach((targetId) => {
      const targetLabel = nodeLabels.get(targetId) ?? targetId;
      const connLabel = new Gtk.Label({ label: `  \u2514\u2500\u2500\u2192 ${targetLabel}`, xalign: 0 });
      connLabel.get_style_context().add_class('dim-label');
      card.append(connLabel);
    });

    const gesture = new Gtk.GestureClick();
    gesture.connect('released', () => onSelectNode?.(node.id));
    card.add_controller(gesture);

    box.append(card);
  });

  if (nodes.length === 0) {
    const empty = new Gtk.Label({ label: '(empty workflow)' });
    empty.get_style_context().add_class('dim-label');
    box.append(empty);
  }

  // Add node button
  const addBtn = new Gtk.Button({ label: '+ Add Node' });
  addBtn.get_style_context().add_class('flat');
  if (onAddNode) addBtn.connect('clicked', onAddNode);
  box.append(addBtn);

  return box;
}
